### ISSUE TITLE
Octant physics negatively impact gridmap performance

### ISSUE THREAD

**Issue by Ando** [godotengine/godot#28801]
**Godot version:** 3.1.1 mono (64 bit)

**Issue description:**
When using the gridmap to render thousands of cells (100.000 cells, 100 different more complex meshes), the framerate drops significantly.
After some investigation, I found out that this is caused by the physics tick. We confirmed this hypothesis by disabling all collision layers.

Additionally, the more cells are filled, the slower it gets to change cells. When we increased the octant size, this had a positive effect on this issue.

Is it possible to disable the physics altogether? Or providing a way to access those physics objects by making octants available through the gridmap?

If not, is there a reason why it is implemented as such? Would there be a problem if we attempted to fix this issue (no C++ experts here though :) )?

**Steps to reproduce:**
Create a gridmap and fill 100.000 cells with around 20 different meshes -a little more complex than a cube-. This should drop the framerate significantly.
When disabling the collision layers, it should be back to normal.

The game will also take a lot longer when changing a cell on a gridmap with 100.000 filled cells than with a gridmap of 1.000.

Example of the meshes used:
![](https://i.gyazo.com/bc0e798bcf311ddc1bb21efc1c4f64ea.png)

**Comment by DoctorWhoof** [comment]
Same here. Splitting the map into octants should improve the performance, but something is wrong.

I have a simple 50x1x50 grid, with simple static box colliders, and about 30 characters checking for collisions with the GridMap (all of them also using box colliders and kinematic bodies). Each tile dimension is about (1.0, 1.5, 1.5).

Setting octant size to 8 dramatically increases the cpu usage and the frame rate drops to a crawl, less than 5 fps on a laptop.

Bringing the octant size to 1 reduces the cpu use enough for 60fps playback on desktop (although still very high, around 70% to 80%), which makes no sense to me - shouldn't a size of 1 basically force collision checks with all tiles, regardless of which octant they belong to? Or am I missing something here?

According to the profiler, the time spent executing any gdscripts is minimal, and pretty much the entire update loop is dominated by the physics process.

I was hoping to figure this one out before moving on to larger maps!

**Comment by Calinou** [comment]
@Ando @DoctorWhoof Does this occur if you set **Physics > 3d > Physics Engine** to **GodotPhysics** in the Project Settings? The default is to use Bullet, on which similar performance issues have been reported; see #26054.

**Comment by KoBeWi** [comment]
Can anyone still reproduce this bug in Godot 3.2.3 or any later release?

If yes, please ensure that an up-to-date Minimal Reproduction Project (MRP) is included in this report (a MRP is a zipped Godot project with the minimal elements necessary to reliably trigger the bug). You can upload ZIP files in an issue comment with a drag and drop.

**Comment by akoutsoulelos** [comment]
I was facing the same problem in Godot 3.2.1 (X11/64bit) and upgraded to Godot 3.2.3 (X11/64bit) to test the new version.
The problem still remains, the FPS is 1 max. 5 when rendering a gridmap, either using GodotPhysics or Bullet, either using a meshlib with collision shapes or not. I tried probing with different octant sizes, even creating multiple gridmaps (to help culling) but the results were all the same. The map size is 64x64.

I'm using a Lenovo IdeaPad [Core i3-6006U@2.00GHz / 4GB RAM / Intel HD Graphics 520] with LinuxMint, running Godot on both GLES2 and GLES3.

I have created a MRP, but due to meshlib the zip size is 28MB. Is there any other way to upload the file?

**Comment by Calinou** [comment]
> I have created a MRP, but due to meshlib the zip size is 28MB. Is there any other way to upload the file?

You can create a public repository on GitHub. You should be able to push it as long as no individual file in the repository is larger than 20 MB.

**Comment by akoutsoulelos** [comment]
OK, here it is - used my own resources: http://www.bitcraft-labs.gr/mrp.[godot].7z

**Comment by belzecue** [comment]
I'm going to take a punt here and say mesh density is your issue. 200 million verts per frame is... busy!

![Screenshot at 2021-06-07 18-03-35](https://user-images.githubusercontent.com/1931303/120998490-b9cd0e00-c7ba-11eb-857a-58b1c6f3b10f.png)

Look at your meshlib source .dae file.  You've got 350K verts composing 80 blocks.

![Screenshot at 2021-06-07 18-04-51](https://user-images.githubusercontent.com/1931303/120998675-e2ed9e80-c7ba-11eb-9acb-a73640407391.png)

![Screenshot at 2021-06-07 18-10-01](https://user-images.githubusercontent.com/1931303/120999365-9ce50a80-c7bb-11eb-86b1-afe15e174865.png)

**Comment by akoutsoulelos** [comment]
That is really surprising: I was just opening again the project to try a different approach; and simultaneously I recieved a notification for your comment....
OK, the meshes were created in VoxelShop, went through Blender to be converted to Collida (.dae) format before being imported in Godot. I believed that the vertices could count up to as many as the color polygons, but it seems I was wrong; I never really checked the wireframe in Blender to see the mess that is pictured in the screenshots above.
Now I was starting to create the cubes from schratch in blender to see if the problem remains - though I am going to change the artistic style, as I decided to drop voxel style for now.
I was planning to comment here in the next few days, and I'm still going to, when I finish the basics in the new approach.
But I think it is clear that the vertices number is the problem...

**Comment by belzecue** [comment]
I suspect there is still an underlying problem, because OP was only dealing in simple geometry.  I've just started my first big gridmap project, so I I'll contribute more to this investigation and discussion over the coming months.

**Comment by akoutsoulelos** [comment]
I remodelled the meshes in Blender, now they are plain cubes (8 verteces) with a simple material.
In Godot, I added a trimesh collision body (StaticBody + CollisionShape) and converted it to a mesh library.
The map I created is 128x128 tiles big. The rate is:
- not more than 20 fps when octant size is set to 8
- not more than 24 fps when octant size is set to 16
The project never loads when octant size is set to less than 8 or above 32.

**Comment by belzecue** [comment]
I've completed the foundation of my prototype that uses a gridmap with 1-unit cubes running on Godot stable 3.3.2 and the default octant value of 8.   I experienced no issues with gridmap performance, however I will note the following optimizations that keep total drawn vertex count low on my little NUC7PJ running Linux:

1. In constructing my 3D maze, I use a mesh library of around 64 meshes that comprise every combination of 6-sided cube for possible visible faces.  Only one of those meshes has all 8 vertices.  Most are just one or a few faces/sides, so the vertex count is less than 8.  Each mesh "cube" has a trimesh collisionbody.
2. I use a short fairly short draw distance, and use fog to give a nice falloff.  This ensures I'm only drawing around max 10K vertices at any time, making a non-issue of lack of occlusion culling.
3. My maze "volume" is any power of two XYZ, e.g. 32 x 32 x 32.  Because of the shortened draw distance, any restraint on performance is due to stuff other than vertex count.  I can construct *BIG* 3D mazes from my gridmap and see no FPS drop.

Here's a video from an early test build before I improved various things.  It's probably only using a smallish volume size used here.  I'll record a newer video soon that gives a better look at performance.

https://www.youtube.com/watch?v=npfMQfLf-G0

So, to sum up my experience to date with gridmaps, I have not been able to trigger any gridmap performance problems yet -- but that could be due to my optimizations.  As I continue to develop this 3D maze project I will push things further to see if I can trigger any of the issues reported in this ticket.

**Comment by belzecue** [comment]
Also, @akoutsoulelos, could you make available your new repro project so I can take a look, thx.

**Comment by belzecue** [comment]
@akoutsoulelos, sorry to spam today, but one thing I just remembered about my test project: I set my collision layers so that my gridmap cube meshes don't collide with each other.  Is that something you've set too in your project?

**Comment by akoutsoulelos** [comment]
@belzecue I have made no optimizations at all, not even resetting collision layers; I just used the gridmap to create a map with everything on its default values.

But now, as far as I understand from how you have set up your project, it seems that gridmap is just a way to align meshes in a 3d grid, and nothing more; at least I would have expected that the engine would not check collisions between 3d tiles/cubes by default. Having the opposite in mind made me think that no optimizations would be needed to create a simple map and that's why I wondered about the fps rate when nothing more was happening than just rendering and moving the camera around the map.

So, it seems that there is no real bug/problem, but it is more the way the gridmap works. In the next days, I will revise my approach and I'll let you know. In any case I'll create an MRP and post it here.

**Comment by akoutsoulelos** [comment]
@belzecue If I understood well you meant to change the collision layers in the static body of each mesh in the meshlibrary, right? 'Cause I did that (set the collision mask to 20 for each mesh in the meshlibrary, leaving the layer to 1) but nothing changed.

**Comment by belzecue** [comment]
@akoutsoulelos, you are right, that setting has no relevence to this issue.  Having played around a bit with my project, I can now see how octant size affects vertices and draw calls: higher octant means more vertices rendered but less draw calls; lower octant size means less vertices rendered but more draw calls.  That makes sense: a higher octant value results in fewer but bigger cell chunks.  If you are getting a higher framerate with a higher octant size, it suggests you are using less draw calls but rendering more vertices.

**Comment by akoutsoulelos** [comment]
@belzecue I'm a little confused; OP described that disabling collision layers restores the fps rate to its normal state, indicating that the problem laid at the physics (collisions). But, when I remove collision bodies from the meshlibrary my results are the same, with no increase of the fps rate.
I had already checked that before, but I was not aware of the vertices mesh you pointed out and I thought that this would the problem, but alas not!

Here is the new MRP: [mrp.[godot].zip](https://github.com/godotengine/godot/files/6658457/mrp.godot.zip)

I will also try it in a different PC, but I think that it will not make any difference.

**Comment by belzecue** [comment]
I took a look at your latest MRP.  I am getting around low 20s FPS when the screen is mostly filled.  Mind you, this is on my little NUC with its graphically puny Intel UHD Graphics 605 GPU.  What GPU are you using?

Here are my stats when the screen is busiest:

![Screenshot at 2021-06-16 17-27-40](https://user-images.githubusercontent.com/1931303/122195984-68fb9a80-cec9-11eb-8a5a-2ddb0266b516.png)

As you can see, it's pushing a lot of vertices (1.25M) and a lot of draw calls per frame.

First thing I did was go in and change your material from Cull Mode: Disabled to Cull Mode: Back, so that it wasn't rendering backfaces, but this gave no improvement in the stats.

Next, I checked your individual meshes, and they can be optimized further, as you have duplicate vertices.  For example:

Original: 26 vertices
![Screenshot at 2021-06-16 17-58-50](https://user-images.githubusercontent.com/1931303/122199810-073d2f80-cecd-11eb-9c7d-86471a0fde16.png)

After merging coexisting vertices in Blender: 8 vertices
![Screenshot at 2021-06-16 18-01-18](https://user-images.githubusercontent.com/1931303/122199936-2340d100-cecd-11eb-8ae4-fe74b8a7a0e1.png)

So that would bring the overall vertex count down from 1.25M to perhaps 0.5M, assuming similar savings on the other meshes after cleaning.

Next, shadows.  There's a reason games only display full shadows in a bubble around the player.  When I switched off shadowing from your main light, my FPS jumped to 40.  So, routine management of shadows and shadow culling will boost performance.  Realtime 3D lighting is expensive, and it should be baked if possible.

![Screenshot at 2021-06-16 18-07-04](https://user-images.githubusercontent.com/1931303/122200653-d27da800-cecd-11eb-9648-b5730d5a895b.png)

The last obvious optimization is culling.  Since we don't have the luxury of occlusion culling in Godot 3.3, we either have to make our own or rely on camera draw distance and fog/fading to hide the boundary.  In your project, you've got your camera Far distance set to 8192, so basically it will draw all the way to the horizon.  And if you're not managing LOD then this is a lot of extra work for no good reason since we cannot see those high fidelity meshes as displayed in a few pixels. (Can't wait for Juan's automatic LOD culling in Godot 4!!)

By shortening the camera draw distance and using fog (and with shadows still switched off), I got back to 60 FPS.  (This is, of course, not a practical solution when you want the entire thing on screen.)

![Screenshot at 2021-06-16 18-22-59](https://user-images.githubusercontent.com/1931303/122203014-32754e00-ced0-11eb-89cc-bb96ca8b3291.png)

Now, all these standard optimisations are fine and dandy, but why am I going so off topic with this reply that still does not directly address the original issue of gridmaps and poor physics performance.  I'm including this optimization stuff in response to your observation that "when I remove collision bodies from the meshlibrary my results are the same, with no increase of the fps rate."  If in your case even with no physics active on your gridmap you get the same low FPS then it means one of two things: either your problem is the standard optimization headache of managing frame content ("polygon budget, lighting budget, texture budget" etc.), or there's a performance problem with gridmaps caused by something *other* that physics.  For our MRPs, if we can eliminate the "not optimized" cause, that leaves gridmap functionality itself as the likely cause.

I've yet to start playing with physics in my gridmap project.  That will arrive in time, and hopefully I can then contribute more meaningfully to the original topic here.

**Comment by akoutsoulelos** [comment]
@belzecue I'm sorry for the late response, but I was a little busy in my job.
First of all, I'm really thankful for your effort and time taken to look in the MRP.
I'm developing on a laptop with i3-6006U@2.00GHz with integraded Intel 520 graphics chip. The idea is that if it can here, it can run everywhere.
I'm looking into your tips, but before I post anything further, I have two questions:
(a) How did you count so many vertices in the meshes? In the original blender file each mesh has 8 vertices, so has in the .dae file when imported to blender (the only difference is that instead of 6 faces, I have 12 in .dae file) - I specifically checked on that, after your first post...
(b) The statistics displayed in the screenshots are with an addon (asset) or is a custom module of yours? It is very usefull to quickly spot basic problems.

**Comment by belzecue** [comment]
@akoutsoulelos, I exported a couple of your meshes out of Godot to .obj and inspected them in Blender.  I guess it's possible that Godot adds vertices when importing a mesh, depending on import settings.  That would explain an imported mesh of, say, 8 vertices getting exported with 12 -- if that's what is happening, which I have not looked into.  And yes, the stats display is a terrific free plug-in by Hungry Proton: https://github.com/HungryProton/monitor_overlay

I suspected you had an Intel onboard graphics chip when I saw I was getting similar framerates to you.  As I said, I run Linux on a tiny little Intel NUC box with no fancy hardware in it, and it has Intel UHD Graphics 605 GPU.  I cannot stress enough how terrible Intel onboard graphics are with OpenGL and Linux.  I think the situation is better on Windows, with better Intel drivers.  Googling just now, here's a reliable comparison: https://www.phoronix.com/scan.php?page=article&item=win10-linux-glcoffee&num=3

That one is from around 2017, so it's not up to date, and it suggest Intel integrated graphics performance on Windows is only a little ahead of Linux, but in my experience I've always seen issues with Intel GPUs and Linux.  On my current Linux NUC, I get weird stuttering all the time when running a build, despite the stats telling me I'm getting a smooth 60 FPS.  When I run the build on Windows, the framerate is smooth and constant.  Anway, that's just me.  I've learned that developing on an Intel GPU is fine, but to test a build properly I need to run it on a device with Windows, or on Linux but with a non-Intel GPU.

However, if I've picked the right GPU for you in the comparison below, I would expect you to be getting double the performance of my Intel integrated GPU:

https://www.videocardbenchmark.net/compare/Intel-HD-520-vs-Intel-UHD-605/3255vs3861

"The idea is that if it can here, it can run everywhere." -- exactly my philosophy too.

**Comment by akoutsoulelos** [comment]
@belzecue Thanks for the plugin - it is really usefull!!
I'm developing also on Linux. But I think that the article you cited is a bit outdated. Intel has a very good support for their drivers on linux, since they are open-sourced.
I looked more thoroughly the vertices number, since I was using the standard collada exporter (since better collada is outdated) and firstly I thought that this may the problem. But the exported .dae file has the minimum number of verices - in my case 8 for each tile. So I exported the tileset using godot scene exporter (that exports in binary format), but the problem persisted. So, I searched the internet, and found [this](https://godotforums.org/discussion/25906/is-there-any-way-of-preventing-godot-splitting-vertices-on-mesh-import). It seems that the renderer itself creates the additional vertices to store the UV map information. So, I will optimaze the tileset, to the minimum needed vertices (propably using mostly planes than cubes) and we'll see.
Now about the rest hints:
- I had forgotten to reset the CullingMode property to 'Back' (I was checking all the choices to see if it made any difference - I thought that checking for the front/back faces may have caused the increased CPU usage);
- You're right about the light - I have already fixed that and the framerate has raised to about 40-45; I was really exagerating about how much shadowing light sources I had (environmental, directional etc.);
- The camera distance is something intensional, but it seems that I have to restrict how much freely the camera moves, in order to keep the rate in viable margins - I have to rethink the perspective.
- Fog is something that I really do not want, since it's going to be a city-building game; I was planning to use fog as a wheather condition (such as rain, snow etc.)

So, I'm going to apply all these changes and post again with the results.

**Comment by lawnjelly** [comment]
I had a little look at the MRP posted by @akoutsoulelos and it does seem as though the main problem is the number of verts involved. When converted to a single mesh, there are 3 million. If you call `make_baked_meshes()` you get some improvement.

For your flat areas you are drawing the whole 3d shape of a cell, when only the top quad needs showing. There also presumably no sharing going on between verts of adjacent cells.

Out of interest you can convert them to regular meshes something like this (if you mark the gridmap as `use_in_baked_light`, it seems to only bake them if this is set):
```
func convert():
	var gridmap = $RoomList/Room_A/GridMap
	gridmap.make_baked_meshes()
	var arr = gridmap.get_bake_meshes()
	
	var num_meshes = arr.size() / 2
	for i in range (num_meshes):
		var ref_mesh = arr[i * 2]
		var mesh_xform = arr[(i * 2) + 1]
		var mi = MeshInstance.new()
		mi.mesh = ref_mesh
		mi.transform = mesh_xform
		
		gridmap.get_parent().add_child(mi)
		
	gridmap.queue_free()
```
You might be able to run some kind of post processing step over this, if this is for a static map, and remove all the unnecessary polys and duplicate verts.

**Comment by akoutsoulelos** [comment]
@lawnjelly, yes the first MRP was a real mess, since I had not checked how objects were exported from blender. So I recreated them to have the minimum possible verteces in an second MRP
@belzecue, since there was a bottleneck in FPS with gridmap, I tried to create the landscape with shaders. It was (and still is) quite tricky. First I created a big plane mesh, subdivided it and applied the heightmap to all the intermediate verteces. But there was no culling and the whole plane was drawn all the time even if a small part was shown (which makes sence since it is one big mesh, and when even one vertex is shown the whole plane must be drawn). So I broke it down creating multiple smaller plane meshes to reduce the verteces, which actually worked. But now I'm stuck with making the shader material unique to each instance of the smaller meshes (each shader must create a different part of the heightmap... The only thing I have not tried yet is to load the shader as the MeshInstance material and not as the mesh material.
Nevertheless, I believe that even at the current stage some interesting conclusions may be deducted, so here is the updated [mrp.[godot].zip](https://github.com/godotengine/godot/files/6754042/mrp.godot.zip). I have created some exported variables for the Game spatial to make it easier to swap between gridmap/shader solutions.
So it seems that both Gridmap and Shaders use excatly the same number of verteces to draw the landscape, but since the shader is run completly on the GPU it gives a slight better FPS rate, but not an important one (Gridmap: ~50 FPS | Shader: ~ 55 FPS), at least for my configuration.
